### PR TITLE
Fix/fix sdl crash on boot

### LINKED ISSUE #2569: SmartDeviceLink crashes on boot.
### Bug Report
SmartDeviceLink crashes on boot.

##### Reproduction Steps
1. Simulate ignition OFF.
(Send from HMI send notification OnExitAllApplications(IGNITION_OFF))
2. Wait 2 minutes.
3. Simulate ignition ON. (start SDL)
4. Wait between 1 and 16 seconds (varying in each iteration).
5. Simulate gear change to R
(Send from HMI send request "VehicleInfo" {prndl="REVERSE"})
6. Wait between 6 and 12 seconds (varying in each iteration).
7. Simulate gear change to D
(Send from HMI send request "VehicleInfo" {prndl="DRIVE"})
8. After 10 seconds, simulate gear change to P.
(Send from HMI send request "VehicleInfo" {prndl="PARK"})
9. Wait 2 minutes.
Repeat 1 - 9 for several times.

##### Expected Behavior
No catastrophic failures. All functionality verified.

##### Observed Behavior
SmartDeviceLink crashes on boot.

##### OS & Version Information
* OS/Version: Ubuntu 16.04
* SDL Core Version: develop

### PR DESCRIPTION
**Pull request by IGapchuk**
Fixes #2569

This PR is ready for review.

### Risk
This PR makes no API changes.

### Testing Plan
Manual testing.

### Summary
The root cause of the core crash could be related to lazy initialization of the HMI/Mobile API classes.
These classes are initializated on demand when some of threads are trying to get the object of that class first time. Most likely at SDL start could be a situation when HMI/Mobile API classes are constructing in one thread
and at that moment another thread is trying to get access to the inner data of not-yet-fully-constructed object and this causes a core crash with ~ISchemaItem() calls in stack trace.

In this PR initialization of HMI/Mobile/V4 classes were moved to AM ctor and were removed useless getters.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/src/rpc_handler_impl.cc [hunk 1]
```diff
@@ -45,6 +45,8 @@ RPCHandlerImpl::RPCHandlerImpl(ApplicationManager& app_manager)
     , messages_from_hmi_("AM FromHMI", this)
     , hmi_so_factory_(hmi_apis::HMI_API())
     , mobile_so_factory_(mobile_apis::MOBILE_API())
+    , v4_protocol_so_factory_(
+          ns_smart_device_link_rpc::V1::v4_protocol_v1_2_no_extra())
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/rpc_handler_impl.cc]: @IGapchuk as I could see `hmi_so_factory_`, `mobile_so_factory_`, `v4_protocol_so_factory_` are declared as a class members. I think they could be default-constructed just like this:
```
, hmi_so_factory_()
, mobile_so_factory_()
, v4_protocol_so_factory_()
```
Can't see any reason to create them from the same copy

File: src/components/application_manager/src/rpc_handler_impl.cc [hunk 2]
```diff
@@ -372,7 +374,7 @@ bool RPCHandlerImpl::ConvertMessageToSO(
 
           smart_objects::SmartObjectSPtr msg_to_send =
               std::make_shared<smart_objects::SmartObject>(output);
-          v1_shema.attachSchema(*msg_to_send, false);
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/src/rpc_handler_impl.cc]: @IGapchuk what is `v1_shema`? If you replaced it with class member then it most likely become unused now. Shouldn't you remove it entirely?

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h [hunk 3]
```diff
@@ -172,6 +172,8 @@ class RPCHandlerImpl : public RPCHandler,
 
   hmi_apis::HMI_API hmi_so_factory_;
   mobile_apis::MOBILE_API mobile_so_factory_;
+  ns_smart_device_link_rpc::V1::v4_protocol_v1_2_no_extra
+      v4_protocol_so_factory_;
```
Comment (AKalinich-Luxoft) [inline src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @IGapchuk BTW, do we really need the copies of these huge objects in this rpc_handler? As I could see you could access the same objects in AM using `app_manager_` reference + public getters

Comment (IGapchuk) [reply src/components/application_manager/include/application_manager/rpc_handler_impl.h]: Fixed in 942e36845e9eca9cf6390731be80a8d155e7ecf9.

File: src/components/application_manager/src/rpc_handler_impl.cc (hunk 1 above)
Comment (IGapchuk) [reply src/components/application_manager/src/rpc_handler_impl.cc]: @AKalinich-Luxoft Fixed in 942e36845e9eca9cf6390731be80a8d155e7ecf9.

File: src/components/application_manager/src/rpc_handler_impl.cc (hunk 2 above)
Comment (IGapchuk) [reply src/components/application_manager/src/rpc_handler_impl.cc]: @AKalinich-Luxoft Fixed in 942e36845e9eca9cf6390731be80a8d155e7ecf9.

File: src/components/application_manager/src/rpc_handler_impl.cc [hunk 4]
```diff
@@ -261,7 +259,8 @@ bool RPCHandlerImpl::ConvertMessageToSO(
       }
 
       if (!conversion_result ||
-          !mobile_so_factory().attachSchema(output, true, msg_version) ||
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/rpc_handler_impl.cc]: @IGapchuk is it style change?

Comment (IGapchuk) [reply src/components/application_manager/src/rpc_handler_impl.cc]: @LuxoftAKutsan No, it is not a style change. Before there changes, rps_handler had the hmi_so_factory and mobile_so_factory as members, and he had private getters for there instances. After there changes, rpc_handler has only v4_protocol_so_factory as member. hmi_so_factory and mobile_so_factory exists in application manager only and rpc_handler has access for them by application manager interface.

File: src/components/application_manager/include/application_manager/application_manager_impl.h [hunk 5]
```diff
@@ -1036,6 +1036,18 @@ class ApplicationManagerImpl
    */
   protocol_handler::MajorProtocolVersion SupportedSDLVersion() const OVERRIDE;
 
+  /**
+   * @brief Function returns reference to HMI factory instance.
+   * @return Reference to HMI factory instance.
+   */
```
[the remark this answers is not in the document]
Comment (IGapchuk) [reply src/components/application_manager/include/application_manager/application_manager_impl.h]: @AByzhynar Removed in 5d9953d8f4e086ba5c25482fb8e90decdcd0b238.

**Comment by AByzhynar** [comment]
@IGapchuk Don't forget to squash commits

File: src/components/application_manager/include/application_manager/rpc_handler_impl.h [hunk 6]
```diff
@@ -161,17 +161,15 @@ class RPCHandlerImpl : public RPCHandler,
...
   ApplicationManager& app_manager_;
   // Thread that pumps messages coming from mobile side.
   impl::FromMobileQueue messages_from_mobile_;
   // Thread that pumps messages coming from HMI.
   impl::FromHmiQueue messages_from_hmi_;
 
-  hmi_apis::HMI_API hmi_so_factory_;
-  mobile_apis::MOBILE_API mobile_so_factory_;
+  ns_smart_device_link_rpc::V1::v4_protocol_v1_2_no_extra
+      v4_protocol_so_factory_;
```
Comment (AByzhynar) [inline src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @IGapchuk Absoletely unclear variable name

Comment (IGapchuk) [reply src/components/application_manager/include/application_manager/rpc_handler_impl.h]: @AByzhynar Fixed in c73ed4e0d89041779c761d4fc1f9820fe7c5ae94.

**Comment by jacobkeeler** [comment]
Looks to be a duplicate of https://github.com/smartdevicelink/sdl_core/pull/3581, which is already merged. Closing for now, please reopen if this fix is still relevant.

**Pull request closed without merging**